tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,538 @@
+    ClientRPCResult commitResult =
+        this.kvClient.commit(backOffer, keys, this.startTs, commitTs, regionId);
+    if (!commitResult.isSuccess() && !commitResult.isRetry()) {
+      throw new TiBatchWriteException("commit primary key error: " + commitResult.getError());
+    }
+    if (commitResult.isRetry()) {
+      try {
+        backOffer.doBackOff(
+            BackOffFunction.BackOffFuncType.BoRegionMiss,
+            new GrpcException(
+                String.format("Txn commit primary key failed, regionId=%s", regionId)));
+        // re-split keys and commit again.
+        this.doCommitPrimaryKey(backOffer, key, commitTs);
+      } catch (GrpcException e) {
+        String error =
+            String.format(
+                "Txn commit primary key error, re-split commit failed, regionId=%s", regionId);
+        LOG.error(error);
+        throw new TiBatchWriteException("commit primary key error" + e);
+      }
+    }
+    if (!commitResult.isSuccess()) {
+      String error =
+          String.format(
+              "Txn commit primary key error, regionId=%s, detail=%s",
+              regionId, commitResult.getError());
+      LOG.error(error);
+      throw new TiBatchWriteException("commit primary key error: " + commitResult.getError());
+    }
+
+    // return success
+    return;
+  }
+
+  /**
+   * 2pc - prewrite secondary keys
+   *
+   * @param backOffer
+   * @param primaryKey
+   * @param pairs
+   * @return
+   */
+  public void prewriteSecondaryKeys(
+      BackOffer backOffer, byte[] primaryKey, Iterator<BytePairWrapper> pairs)
+      throws TiBatchWriteException {
+    Iterator<Pair<ByteString, ByteString>> byteStringKeys =
+        new Iterator<Pair<ByteString, ByteString>>() {
+
+          @Override
+          public boolean hasNext() {
+            return pairs.hasNext();
+          }
+
+          @Override
+          public Pair<ByteString, ByteString> next() {
+            BytePairWrapper pair = pairs.next();
+            return new Pair<>(
+                ByteString.copyFrom(pair.getKey()), ByteString.copyFrom(pair.getValue()));
+          }
+        };
+
+    doPrewriteSecondaryKeys(backOffer, ByteString.copyFrom(primaryKey), byteStringKeys);
+  }
+
+  private void doPrewriteSecondaryKeys(
+      BackOffer backOffer, ByteString primaryKey, Iterator<Pair<ByteString, ByteString>> pairs)
+      throws TiBatchWriteException {
+    while (pairs.hasNext()) {
+      ByteString[] keyBytes = new ByteString[WRITE_BUFFER_SIZE];
+      ByteString[] valueBytes = new ByteString[WRITE_BUFFER_SIZE];
+      int size = 0;
+      while (size < WRITE_BUFFER_SIZE && pairs.hasNext()) {
+        Pair<ByteString, ByteString> pair = pairs.next();
+        keyBytes[size] = pair.first;
+        valueBytes[size] = pair.second;
+        size++;
+      }
+      doPrewriteSecondaryKeys(backOffer, primaryKey, keyBytes, valueBytes, size);
+    }
+
+    // return success
+    return;
+  }
+
+  private void doPrewriteSecondaryKeys(
```

Comment:
consider change its name or add comment for this method so we know it does `prewrite` in batches. e.g., `doPrewriteSecondaryKeysInBatches`. The same goes for `doPrewritePrimarykey`.

Reply:
fixed